zebra-chain/src/block/tests/generate.rs

```diff
@@ -1,106 +1,207 @@
     multi_transaction_block(true)
 }
 
-/// Generate a block with one transaction and multiple inputs just above limit
-pub fn oversized_single_transaction_block() -> Block {
-    single_transaction_block(true)
+/// Generate a block with one transaction and multiple transparent inputs just above limit
+///
+/// TODO: add a coinbase height to the returned block
+///       make the returned block stable under round-trip serialization
+pub fn oversized_single_transaction_block_many_inputs() -> Block {
+    single_transaction_block_many_inputs(true)
+}
+
+/// Generate a block with one transaction and multiple transparent outputs just above limit
+///
+/// TODO: add a coinbase height to the returned block
+///       make the returned block stable under round-trip serialization
+pub fn oversized_single_transaction_block_many_outputs() -> Block {
+    single_transaction_block_many_outputs(true)
 }
 
-// Implementation of block generation with multiple transactions
+/// Implementation of block generation with multiple transparent transactions
 fn multi_transaction_block(oversized: bool) -> Block {
     // A dummy transaction
-    let tx = Transaction::zcash_deserialize(&zebra_test::vectors::DUMMY_TX1[..]).unwrap();
+    let (transaction, transaction_bytes) = transaction();
 
     // A block header
-    let header = block_header();
+    let (block_header, block_header_bytes) = block_header();
 
-    // Serialize header
-    let mut data_header = Vec::new();
-    header
-        .zcash_serialize(&mut data_header)
-        .expect("Block header should serialize");
-
-    // Calculate the number of transactions we need
-    let mut max_transactions_in_block =
-        (MAX_BLOCK_BYTES as usize - data_header.len()) / zebra_test::vectors::DUMMY_TX1[..].len();
+    // Calculate the number of transactions we need,
+    // subtracting the bytes used to serialize the expected transaction count.
+    let mut max_transactions_in_block = (usize::try_from(MAX_BLOCK_BYTES).unwrap()
+        - block_header_bytes.len()
+        - MAX_COMPACT_SIZE_BYTES)
+        / transaction_bytes.len();
     if oversized {
         max_transactions_in_block += 1;
     }
 
     // Create transactions to be just below or just above the limit
-    let transactions = std::iter::repeat(Arc::new(tx))
+    let transactions = std::iter::repeat(Arc::new(transaction))
         .take(max_transactions_in_block)
         .collect::<Vec<_>>();
 
     // Add the transactions into a block
-    Block {
-        header,
+    let block = Block {
+        header: block_header,
         transactions,
-    }
+    };
+
+    let serialized_len = block.zcash_serialize_to_vec().unwrap().len();
+    assert_eq!(
+        oversized,
+        serialized_len > MAX_BLOCK_BYTES.try_into().unwrap(),
+        "block is over-sized if requested:\n\
+         oversized: {},\n\
+         serialized_len: {},\n\
+         MAX_BLOCK_BYTES: {},",
+        oversized,
+        serialized_len,
+        MAX_BLOCK_BYTES,
```

Comment:
Suggestion: small tweak for conciseness.

```suggestion
         oversized: {oversized},\n\
         serialized_len: {serialized_len},\n\
         MAX_BLOCK_BYTES: {MAX_BLOCK_BYTES},",
```